frappe/handler.py

```diff
@@ -18,7 +19,16 @@ def handle():
 	cmd = frappe.local.form_dict.cmd
 	data = None
 
-	if cmd!='login':
+	#arg1=&cmd=custom_server_action.{custom_sever_action} parameter will be passed via frappe.form_data,
+	server_action = False
+	if cmd.split(".")[0] == 'custom_server_action':
+		server_action=cmd.split(".",1)[-1]
+	#for mapper related call, the cmd is mapper.xx, method as form parameter
+	if frappe.local.form_dict.get('method','').split('.')[0]=='custom_server_action':
+		server_action =  frappe.local.form_dict.get('method','').split('.',1)[-1]
+	if server_action:
+		data = run_custom_server_action(server_action)
```

Comment:
the frappe-way of writing this is by adding `@whitelist` method

example

Example in `frappe/custom/doctype/custom_server_action/custom_server_action.py`

```
@frappe.whitelist()
def run_custom_server_action(name)
```

then call with

`/api/method/frappe.custom.doctype.custom_server_action.custom_server_action.run_custom_server_action`

This kind of hard-coding in handler.py is not good!

Reply:
moved to custom/doctype/custom_server_action/__init__.py